ec2.tf

```diff
@@ -0,0 +1,181 @@
+      "GroupPendingCapacity",
+      "GroupStandbyCapacity",
+      "GroupTerminatingCapacity",
+      "GroupTotalCapacity",
+      "WarmPoolDesiredCapacity",
+      "WarmPoolWarmedCapacity",
+      "WarmPoolPendingCapacity",
+      "WarmPoolTerminatingCapacity",
+      "WarmPoolTotalCapacity",
+      "GroupAndWarmPoolDesiredCapacity",
+      "GroupAndWarmPoolTotalCapacity",
+    ]
+    service_linked_role_arn              = ""
+    metadata_http_endpoint_enabled       = true
+    metadata_http_put_response_hop_limit = 2
+    metadata_http_tokens_required        = true
+    metadata_http_protocol_ipv6_enabled  = false
+    tag_specifications_resource_types = [
+      "instance",
+      "volume"
+    ]
+    max_instance_lifetime = null
+    capacity_rebalance    = false
+    warm_pool             = null
+  }
+
+  ec2_capacity_providers = { for name, provider in local.ec2_capacity_providers_cleanup :
+    name => merge(local.ec2_capacity_provider_default, provider)
+  }
+
+  instance_profile_name = join("", aws_iam_instance_profile.default.*.name)
+}
+
+data "template_cloudinit_config" "default" {
+  for_each      = local.ec2_capacity_providers
+  gzip          = false
+  base64_encode = true
+
+  part {
+    content_type = "text/x-shellscript"
+    content      = <<EOT
+#!/bin/bash
+echo ECS_CLUSTER="${local.cluster_name}" >> /etc/ecs/ecs.config
+echo ECS_ENABLE_CONTAINER_METADATA=true >> /etc/ecs/ecs.config
+echo ECS_POLL_METRICS=true >> /etc/ecs/ecs.config
+echo ECS_ENABLE_SPOT_INSTANCE_DRAINING=${each.value.instance_market_options != null && each.value.mixed_instances_policy != null} >> /etc/ecs/ecs.config
+echo ECS_WARM_POOLS_CHECK=${each.value.warm_pool != null} >> /etc/ecs/ecs.config
+EOT
+  }
+
+  part {
+    content_type = "text/x-shellscript"
+    content      = replace(each.value["user_data"], "#!/bin/bash", "")
+  }
+}
+
+module "autoscale_group" {
+  for_each = local.ec2_capacity_providers
+
+  source  = "cloudposse/ec2-autoscale-group/aws"
+  version = "0.30.1"
+
+  context    = module.this.context
+  attributes = concat(var.attributes, [each.key])
+
+  image_id      = each.value["image_id"]
+  instance_type = each.value["instance_type"]
+
+
+  ## ECS autoscaling group does not have ELB integration, so we use EC2 health check type
+  health_check_type = "EC2"
+  # The Auto Scaling group must have instance protection from scale in enabled to use managed termination protection for a capacity provider,
+  protect_from_scale_in = true
+  # Disable autoscaling rules because scaling would be managed by ECS
+  autoscaling_policies_enabled = false
+  default_alarms_enabled       = false
+
+  iam_instance_profile_name = local.instance_profile_name
+  user_data_base64          = data.template_cloudinit_config.default[each.key].rendered
+  tags                      = merge(module.this.context.tags, { "AmazonECSManaged" : "true" })
+
+
+  instance_initiated_shutdown_behavior = each.value["instance_initiated_shutdown_behavior"]
```

Comment:
since these values are all `optional`, we need to set a default value within `optional()` or this will fail if not provided

Reply:
same for all other values here

Reply:
nevermind, defaults are fined in locals